src/context/key_cursor.rs

```diff
@@ -0,0 +1,280 @@
+                Some(iterator_callback),
+                data_ptr,
+            )
+        };
+
+        // Check if more calls are needed
+        if ret == 0 {
+            self.last_call = true;
+            // we may still have buffered items
+        }
+
+        if stack_slot.buf.is_empty() {
+            // no items were returned, try again
+            None
+        } else {
+            self.next_buffered_item()
+        }
+    }
+
+    fn next_buffered_item(&mut self) -> Option<ScanKeyIteratorItem> {
+        // todo: use different datatype / access pattern for performance
+        Some(self.buf.remove(0))
+    }
+}
+
+/// The callback that is used by [`ScanKeyCursor::foreach`] as argument to `RedisModule_ScanKey`.
+///
+/// The `data` pointer is the closure given to [`ScanKeyCursor::foreach`] and the callback forwards
+/// references to the key, field and value to that closure.
+unsafe extern "C" fn foreach_callback<F: FnMut(&RedisKey, &RedisString, &RedisString)>(
+    key: *mut raw::RedisModuleKey,
+    field: *mut raw::RedisModuleString,
+    value: *mut raw::RedisModuleString,
+    data: *mut c_void,
+) {
+    let ctx = ptr::null_mut();
+    let key = RedisKey::from_raw_parts(ctx, key);
+
+    let field = RedisString::from_redis_module_string(ctx, field);
+    let value = RedisString::from_redis_module_string(ctx, value);
+
+    let callback = unsafe { &mut *(data.cast::<F>()) };
+    callback(&key, &field, &value);
+
+    // we're not the owner of field and value strings
+    field.take();
+    value.take();
+
+    key.take(); // we're not the owner of the key either
+}
+
+/// The callback that is used inside the iterator variant accessible via [`ScanKeyCursor::iter`] for `RedisModule_ScanKey`. 
+/// 
+/// It buffers copies of the field and value strings as the lifetime of them ends when with the call
+/// to `RedisModule_ScanKey` going out of scope.
+/// 
+/// The `data` pointer is a stack slot of type [StackSlot] that is used to pass the data back to the iterator.
+unsafe extern "C" fn iterator_callback(
+    _key: *mut raw::RedisModuleKey,
+    field: *mut raw::RedisModuleString,
+    value: *mut raw::RedisModuleString,
+    data: *mut c_void,
+) {
+    // `data` is a stack slot
+    let slot = data.cast::<StackSlot>();
+    let slot = &mut (*slot);
+
+    // todo: use new-type with refcount handling for better performance, otherwise we have to copy at this point
+    // we know that this callback will be called in a loop scope and that we 
+    // need to store the RedisString longer than the ScanKey invocation but not much
+    // longer and in case of batched results we don't need to store everything in memory
+    // but only the last batch.
+    let field = raw::RedisModule_CreateStringFromString.unwrap()(slot.ctx.get_raw(), field);
+    let value = raw::RedisModule_CreateStringFromString.unwrap()(slot.ctx.get_raw(), value);
+
+    let field = RedisString::from_redis_module_string(slot.ctx.get_raw(), field);
```

Comment:
This will not actually do what you want it to (prevent the strings from being freed) you need to call [`RedisModule_RetainString`](https://github.com/redis/redis/blob/8ad5421502241d6088b701bb4a4262124343345a/src/module.c#L2776-L2810) on the `RedisModuleString` for that ([`safe_clone`](https://docs.rs/redis-module/latest/redis_module/struct.RedisString.html#method.safe_clone) in Rust).